### ISSUE TITLE
Angular Component Testing: provide access to DI

### ISSUE THREAD

**Issue by yjaaidi** [cypress-io/cypress#24429]
### What would you like?

Angular's CT `mount` function should return an `inject` function that allows  DI.

e.g.:
```ts
const {inject} = cy.mount(...);
const service = inject(MyService);
```

where `inject` is nothing more than a wrapper of `TestBed.inject` just to hide the implementation detail which is `TesBed`.
we might also provide a `cy.inject` command, that does something like this to provide a better DX:
```ts
cy.then(() => TestBed.inject(...))
```

What do you think?

cc. @jordanpowell88 

### Why is this needed?

As an advocate for fakes (vs. stubs/spies/mocks), I often end up with stateful services that I might want to access & check manually afterward in my test. But of course, there are way many other use cases like arranging the test etc...

```ts
cy.mount(MyComponent, {providers: [{provide: Cart, useClass: CartFake}]})

cy.contains('Buy').click()

cy.inject(CartFake).invoke('getItems').should('have.length', 1);
```

or

```ts
const { inject } = cy.mount(MyComponent, {providers: [{provide: Cart, useClass: CartFake}]})

cy.contains('Buy').click()

cy.then(() => inject(CartFake).getItems()).should('have.length', 1);
```

cc. @jordanpowell88

**Comment by yjaaidi** [comment]
cc. @rainerhahnekamp

**Comment by rainerhahnekamp** [comment]
This is very much needed.

A popular example might be the injection of the `HttpTestingController`... I know the usage of it is questionable in Cypres. But I think it serves as a good example that most Angular developers are familiar with.

**Comment by jordanpowell88** [comment]
You should be able to reference `TestBed.inject(MyService)` as a workaround for the time being

**Comment by QuantariusRay** [comment]
If this is a component test, I'd be curious why you would be testing your providers. Why would this not just be an integration or end to end test?

**Comment by rainerhahnekamp** [comment]
Quantarius (cool name by the way 👍), it could be that the information of how many items are in `Cart` is not shown in  `MyComponent`. Then the only way to find out is to check the service directly.

But I think Younes just made this example primarily to demonstrate the requirement of the `inject` command.

**Comment by rockindahizzy** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Component-Testing-606e43bbf1b3b20013c142c1/issues/cypress-io/cypress/24429?planning-poker) with Zenhub @amehta265 @astone123 @lmiller1990 @marktnoonan @mike-plummer @warrensplayer @ZachJW34

**Comment by rainerhahnekamp** [comment]
So I was thinking for some time if I could come up with an improved API, but I think @yjaaidi's dedicated `cy.mount` which is only available for Angular component testing, is the best way.

I could then easily do something like this:

```typescript
// slightly modified without fake services

cy.mount(MyComponent)
cy.contains('Buy').click()
cy.inject(Cart).should(cart => expect(card.items).to.have.length(1));
```

This is way more better to read then
```typescript
cy.mount(MyComponent)
cy.contains('Buy').click().then(() => {
  const card = TestBed.inject(Card);
  expect(card.items).to.have.length(1);
})
```

I'd be happy to provide a PR, unless somebody (@ZachJW34, @jordanpowell88)  sees some obstacles.

**Comment by jordanpowell88** [comment]
@rainerhahnekamp Feel free to work on a solution you feel like improves the API. Also for clarification we release `@cypress/angular` as a separate lib on npm for anyone who wants to eject or do something custom with it as apposed to the `cypress/angular` lib (same one) that gets shipped with the binary.

**Comment by lmiller1990** [comment]
Not assigning a vote for sprint planning since scope and API is not yet defined.

**Comment by jonnyleeharris** [comment]
This would be incredibly useful for me. Being able to inject the ApolloTestingController would mean I could mock a GraphQL back end the same way I can in my other tests (outside of Cypress).

**Comment by muhamedkarajic** [comment]
Anyone found some workaround for now?

**Comment by rainerhahnekamp** [comment]
@muhamedkarajic 

Not a workaround, but the "solution" ;) https://github.com/cypress-io/cypress/pull/25505#issuecomment-1481454207

**Comment by lmiller1990** [comment]
I thought we had another issue to track this, I think @jordanpowell88 is working on this.

Proposed API

```js
cy.intercept('GET', '/count', { count: 5 }).then(() => {
    cy.inject(CountService).invokeService('getCount').should('equal', 5)
})
```

**Comment by jordanpowell88** [comment]
Yes we plan to work on implementing a solution soon

**Comment by rainerhahnekamp** [comment]
@jordanpowell88 @lmiller1990

the other issue was that PR https://github.com/cypress-io/cypress/pull/25505. It wasn't accepted though.

So in order to reduce redundancy, you might what to check on that one first?

**Comment by muhamedkarajic** [comment]
> @muhamedkarajic
> 
> Not a workaround, but the "solution" ;) [#25505 (comment)](https://github.com/cypress-io/cypress/pull/25505#issuecomment-1481454207)

Writing on the support team turns out to work ^^. Thanks a bunch I appreciate the quick response.

**Comment by AliF50** [comment]
Can't we get the service in the `.then` block after the `mount` and then have a handle on it throughout the test?

```
let service: Service;

cy.mount(Component, config).then(({ fixture }) => {
  // Now you have a handle on the service
  service = fixture.debugElement.injector.get(Service);
});

// Do stuff with the component to add an item to the service

expect(service.newItems).should('have.length', 1);
```

**Comment by lmiller1990** [comment]
Interesting idea -- I'll see what @jordanpowell88 thinks, since he will likely be the one working on this in our next block of work (next two weeks). It's happening!

**Comment by jordanpowell88** [comment]
@AliF50 That is correct. In fact you can reference the `TestBed` directly from the test as well and do this exact same thing. What we are exploring is a more ergonomic option out of the box

**Comment by mfrieling** [comment]
Just came across this issue while trying to inject a stateful service on application initialization of a standalone app. In my `app.config` I have something like this:

```
import { APP_INITIALIZER } from '@angular/core'
import type { ApplicationConfig } from '@angular/core'

export const appConfig: AppConfiguration = {
  providers: [
    {
      provide: APP_INITIALIZER,
      useFactory: configFactory,
      deps: [ConfigService],
      multi: true,
    },
  ]
}
```

The `ConfigService` has a method to load and parse an `/assets/app.config` file which contains some dynamic parameters which differ between multiple production deployments of the app. The `configFactory` gets the service injected and calls that method on application startup.

In addition the `ConfigService` has a property (`signal<ClientConfig>` but could be a `BehaviorSubject<ClientConfig>` as well) holding the parsed configuration, which will be accessed by several other services.

How is/will the use case of injecting `APP_INITIALIZER`s be covered by the planned implementation? And what is the current schedule for this issue?

**Comment by konstantinschuette** [comment]
> @AliF50 That is correct. In fact you can reference the `TestBed` directly from the test as well and do this exact same thing. What we are exploring is a more ergonomic option out of the box

How's it going?
